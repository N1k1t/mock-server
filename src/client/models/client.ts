import type { IExpectationOperatorContext, TBuildExpectationConfiguration } from '../../expectations';
import type { TMethodsSchema } from '../types';

import { compileExpectationOperators, ICompiledExpectationOperators } from '../helpers';

interface IExpectationHandlerContext<T extends PartialDeep<IExpectationOperatorContext>> {
  $: ICompiledExpectationOperators<T>;
  T: <T>(payload: T) => T extends TFunction<any, any[]> ? TFunction<any, any[]> : T;
}

interface IUnknownExpectationContext {
  incoming: IExpectationOperatorContext['incoming'] & {
    headers: any;
    query: any;
    body: any;
  };

  outgoing?: IExpectationOperatorContext['outgoing'] & {
    headers: any;
    data: any;
  };
}

export class Client {
  constructor(protected methods: TMethodsSchema) {}

  public get ping() {
    return this.methods.ping;
  }

  public get deleteExpectations() {
    return this.methods.deleteExpectations;
  }

  public createExpectation<T extends PartialDeep<IExpectationOperatorContext> = {}>(
    predicate: TBuildExpectationConfiguration<IUnknownExpectationContext> | TFunction<
      TBuildExpectationConfiguration<any>,
      [IExpectationHandlerContext<T>]
    >
  ): ReturnType<TMethodsSchema['createExpectation']> {
    return typeof predicate === 'function'
      ? this.methods.createExpectation(predicate({
        $: compileExpectationOperators<T>(),
        T: (payload: any) => payload,
      }))
      : this.methods.createExpectation(predicate);
  }

  public updateExpectation<T extends Partial<IExpectationOperatorContext>>(
    predicate: { id: string, set: Partial<Omit<TBuildExpectationConfiguration, 'type'>> } | TFunction<
      { id: string, set: Partial<Omit<TBuildExpectationConfiguration<any>, 'type'>> },
      [IExpectationHandlerContext<T>]
    >
  ): ReturnType<TMethodsSchema['updateExpectation']> {
    return typeof predicate === 'function'
      ? this.methods.updateExpectation(predicate({
        $: compileExpectationOperators<T>(),
        T: (payload: any) => payload,
      }))
      : this.methods.updateExpectation(predicate);
  }
}
