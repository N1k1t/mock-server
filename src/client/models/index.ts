export * from './method';
export * from './client';
