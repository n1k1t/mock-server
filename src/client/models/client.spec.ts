import { prepareExpectationBodyToRequest } from '../utils';
import { Expectation } from '../../expectations';
import { Client } from './client';


interface IExpectationContext {
  incoming: {
    query: {
      foo?: 'a' | 'b' | 'c';
      bar: 'a' | 'b' | 'c';
    }

    body: {
      foo: 'a' | 'b' | 'c';
      bar?: ('a' | 'b' | 'c')[];

      baz: {
        foo: 1 | 2 | 3;
        bar?: 1 | 2 | 3;
        baz?: (1 | 2 | 3)[];
      };
    };
  };
}

const buildClient = () => new Client({
  ping: () => Promise.resolve('pong'),

  providersCreate: () => Promise.resolve(null),
  providersDelete: () => Promise.resolve(null),

  expectationsDelete: () => Promise.resolve(null),
  expectationsGroupUpdate: () => Promise.resolve([]),

  expectationsCreate: (configuration) => Promise.resolve(
    prepareExpectationBodyToRequest(
      Object.assign(Expectation.build(configuration).toPlain(), {
        id: '12426262-85a3-4340-969b-272e003722e9',
        name: 'MaliciousCobra',
      })
    )
  ),

  expectationsUpdate: ({ id, set }) => Promise.resolve(
    prepareExpectationBodyToRequest(
      Object.assign(Expectation.build({ ...set, schema: set.schema ?? {} }).toPlain(), { id })
    )
  ),
});

describe('Client.Models.Client', () => {
  it('should create expectation with complex schema as object', async () => {
    const client = buildClient();

    await expect(
      client.createExpectation({
        schema: {
          request: {
            $and: [
              { $has: { $location: 'path', $value: '/foo' } },
              { $has: { $location: 'path', $valueAnyOf: ['/foo', '/foo/bar'] } },
              { $has: { $location: 'path', $match: '/foo/*' } },
              { $has: { $location: 'path', $matchAnyOf: ['/foo/*', '*/bar'] } },
              { $has: { $location: 'path', $regExp: /^\/foo/ } },
              { $has: { $location: 'path', $regExpAnyOf: [/^\/foo$/, /^\/foo\/bar$/] } },
              { $has: { $location: 'path', $exec: (payload) => payload.includes('/foo') } },

              { $has: { $location: 'path', $path: '/foo', $value: '/foo' } },
              { $has: { $location: 'path', $path: '/foo', $valueAnyOf: ['/foo', '/foo/bar'] } },
              { $has: { $location: 'path', $path: '/foo', $match: '/foo/*' } },
              { $has: { $location: 'path', $path: '/foo', $matchAnyOf: ['/foo/*', '*/bar'] } },
              { $has: { $location: 'path', $path: '/foo', $regExp: /^\/foo/ } },
              { $has: { $location: 'path', $path: '/foo', $regExpAnyOf: [/^\/foo$/, /^\/foo\/bar$/] } },
              { $has: { $location: 'path', $path: '/foo', $exec: (payload) => payload.includes('/foo') } },

              { $has: { $location: 'incoming.query' } },

              { $has: { $location: 'incoming.query', $value: { foo: 'a' } } },
              { $has: { $location: 'incoming.query', $valueAnyOf: [{ foo: 'a' }, { foo: 'b' }] } },
              { $has: { $location: 'incoming.query', $match: { foo: 'a' } } },
              { $has: { $location: 'incoming.query', $matchAnyOf: [{ foo: 'a' }, { foo: 'b' }] } },
              { $has: { $location: 'incoming.query', $exec: (payload) => payload?.foo === 'a' } },

              { $has: { $location: 'incoming.query', $path: 'foo', $value: 'a' } },
              { $has: { $location: 'incoming.query', $path: 'foo', $valueAnyOf: ['a', 'b'] } },
              { $has: { $location: 'incoming.query', $path: 'foo', $match: 'a' } },
              { $has: { $location: 'incoming.query', $path: 'foo', $matchAnyOf: ['a', 'b'] } },
              { $has: { $location: 'incoming.query', $path: 'foo', $regExp: /a|b/ } },
              { $has: { $location: 'incoming.query', $path: 'foo', $regExpAnyOf: [/a/i, /b/i] } },
              { $has: { $location: 'incoming.query', $path: 'foo', $exec: (payload) => payload === 'a' } },

              { $has: { $location: 'incoming.data' } },

              { $has: { $location: 'incoming.data', $value: { foo: 'a', bar: ['a'] } } },
              { $has: { $location: 'incoming.data', $valueAnyOf: [{ foo: 'a', bar: ['a'] }, { bar: ['b'] }, { baz: {} }] } },
              { $has: { $location: 'incoming.data', $match: { foo: 'a', baz: { foo: 1 } } } },
              { $has: { $location: 'incoming.data', $matchAnyOf: [{ baz: {} }, { bar: [] }] } },
              { $has: { $location: 'incoming.data', $exec: (payload) => payload?.baz?.foo === 1 } },

              { $has: { $location: 'incoming.data', $path: 'baz', $value: { foo: 1, bar: 1 } } },
              { $has: { $location: 'incoming.data', $path: 'baz', $valueAnyOf: [{ foo: 1 }, { bar: 1 }] } },
              { $has: { $location: 'incoming.data', $path: 'baz', $match: { foo: 1 } } },
              { $has: { $location: 'incoming.data', $path: 'baz', $matchAnyOf: [{}, { foo: '1' }] } },
              { $has: { $location: 'incoming.data', $path: 'baz', $exec: (payload) => payload?.foo === 1 } },

              { $has: { $location: 'incoming.data', $path: 'baz.baz', $value: [1] } },
              { $has: { $location: 'incoming.data', $path: 'baz.baz', $valueAnyOf: [[1], [1, 2]] } },
              { $has: { $location: 'incoming.data', $path: 'baz.baz', $match: '*' } },
              { $has: { $location: 'incoming.data', $path: 'baz.baz', $matchAnyOf: ['1', '2'] } },
              { $has: { $location: 'incoming.data', $path: 'baz.baz', $regExp: /1/ } },
              { $has: { $location: 'incoming.data', $path: 'baz.baz', $regExpAnyOf: [/1/, /2/] } },
              { $has: { $location: 'incoming.data', $path: 'baz.baz', $exec: (payload) => payload === 1 } },
            ],
          },

          response: {
            $or: [
              {
                $switch: {
                  $location: 'outgoing.status',
                  $cases: {
                    200: {
                      $and: [
                        { $set: { $location: 'outgoing.data', $value: {} } },
                        { $set: { $location: 'outgoing.data', $path: 'foo', $value: 'a' } },

                        { $set: { $location: 'outgoing.headers', $exec: () => ({}) } },
                        { $set: { $location: 'outgoing.headers', $path: 'foo', $value: 'bar' } },
                      ],
                    },
                  },

                  $default: { $set: { $location: 'outgoing.data', $value: {} } },
                },
              },
              {
                $switch: {
                  $location: 'outgoing.data',
                  $path: 'baz.bar',

                  $cases: {
                    1: {
                      $and: [
                        { $merge: { $location: 'outgoing.data', $value: {} } },
                        { $remove: { $location: 'outgoing.data', $path: 'foo' } },
                      ],
                    },
                  },
                },
              },
              {
                $switch: {
                  $location: 'outgoing.headers',
                  $exec: (payload) => payload.foo,

                  $cases: {
                    'a': {
                      $and: [{
                        $exec: ({ context }) => {
                          if (context.incoming.data?.foo === 'a') {
                            context.outgoing!.status = 200;
                          }
                        },
                      }],
                    },
                  },
                },
              },
            ],
          },
        },
      })
    ).resolves.toMatchSnapshot();
  });

  it('should create expectation with complex schema as function', async () => {
    const client = buildClient();

    await expect(
      client.createExpectation<IExpectationContext>(({ $ }) => ({
        schema: {
          request: $.and([
            $.has('path', { $value: '/foo' }),
            $.has('path', { $valueAnyOf: ['/foo', '/foo/bar'] }),
            $.has('path', { $match: '/foo/*' }),
            $.has('path', { $matchAnyOf: ['/foo/*', '*/bar'] }),
            $.has('path', { $regExp: /^\/foo/ }),
            $.has('path', { $regExpAnyOf: [/^\/foo$/, /^\/foo\/bar$/] }),
            $.has('path', { $exec: (payload) => payload.includes('/foo') }),

            $.has('path', '$path', '/foo', { $value: '/foo' }),
            $.has('path', '$path', '/foo', { $valueAnyOf: ['/foo', '/foo/bar'] }),
            $.has('path', '$path', '/foo', { $match: '/foo/*' }),
            $.has('path', '$path', '/foo', { $matchAnyOf: ['/foo/*', '*/bar'] }),
            $.has('path', '$path', '/foo', { $regExp: /^\/foo/ }),
            $.has('path', '$path', '/foo', { $regExpAnyOf: [/^\/foo$/, /^\/foo\/bar$/] }),
            $.has('path', '$path', '/foo', { $exec: (payload) => payload.includes('/foo') }),

            $.has('incoming.query', {}),
            $.has('incoming.query', '$path', 'foo'),

            $.has('incoming.query', { $value: { foo: 'a', bar: 'a' } }),
            $.has('incoming.query', { $valueAnyOf: [{ foo: 'a', bar: 'a' }, { foo: 'b', bar: 'b' }] }),
            $.has('incoming.query', { $match: { foo: 'a' } }),
            $.has('incoming.query', { $matchAnyOf: [{ foo: 'a' }, { foo: 'b' }] }),
            $.has('incoming.query', { $exec: (payload) => payload?.foo === 'a' }),

            $.has('incoming.query', '$path', 'foo', { $value: 'a' }),
            $.has('incoming.query', '$path', 'foo', { $valueAnyOf: ['a', 'b'] }),
            $.has('incoming.query', '$path', 'foo', { $match: 'a' }),
            $.has('incoming.query', '$path', 'foo', { $matchAnyOf: ['a', 'b'] }),
            $.has('incoming.query', '$path', 'foo', { $regExp: /a|b/ }),
            $.has('incoming.query', '$path', 'foo', { $regExpAnyOf: [/a/i, /b/i] }),
            $.has('incoming.query', '$path', 'foo', { $exec: (payload) => payload === 'a' }),

            $.has('incoming.data', {}),

            $.has('incoming.data', { $value: { foo: 'a', bar: ['a'] } }),
            $.has('incoming.data', { $valueAnyOf: [{ foo: 'a', bar: ['a'] }, { bar: ['b'] }, { baz: {} }] }),
            $.has('incoming.data', { $match: { foo: 'a', baz: { foo: 1 } } }),
            $.has('incoming.data', { $matchAnyOf: [{ baz: {} }, { bar: [] }] }),
            $.has('incoming.data', { $exec: (payload) => payload?.baz?.foo === 1 }),

            $.has('incoming.data', '$path', 'baz', { $value: { foo: 1, bar: 1 } }),
            $.has('incoming.data', '$path', 'baz', { $valueAnyOf: [{ foo: 1 }, { foo: 1, bar: 1 }] }),
            $.has('incoming.data', '$path', 'baz', { $match: { foo: 1 } }),
            $.has('incoming.data', '$path', 'baz', { $matchAnyOf: [{}, { foo: 1 }] }),
            $.has('incoming.data', '$path', 'baz', { $exec: (payload) => payload?.foo === 1 }),

            $.has('incoming.data', '$jsonPath', '123.123.213', {
              $match: {},
            }),

            $.has('incoming.data', '$path', 'baz.baz', { $value: [1] }),
            $.has('incoming.data', '$path', 'baz.baz', { $valueAnyOf: [[1], [1, 2]] }),
            $.has('incoming.data', '$path', 'baz.baz.0', { $match: '*' }),
            $.has('incoming.data', '$path', 'baz.baz.0', { $matchAnyOf: ['1', '2'] }),
            $.has('incoming.data', '$path', 'baz.baz', { $regExp: /1/ }),
            $.has('incoming.data', '$path', 'baz.baz', { $regExpAnyOf: [/1/, /2/] }),
            $.has('incoming.data', '$path', 'baz.baz', { $exec: (payload) => payload?.[0] === 1 }),

            $.switch('incoming.query', '$path', 'foo', {
              $cases: {
                'a': $.and([])
              },
            }),
          ]),

          response: $.or([
            $.switch('outgoing.status', {
              $cases: {
                200: $.and([
                  $.set('outgoing.data', { $value: {} }),
                  $.set('outgoing.data', '$path', 'foo', { $value: 'a' }),
                  $.set('outgoing.headers', { $exec: () => ({}) }),
                  $.set('outgoing.headers', '$path', 'foo', { $value: 'bar' }),
                ]),
              },

              $default: $.set('outgoing.data', { $value: {} }),
            }),

            $.switch('incoming.data', '$path', 'baz.bar', {
              $cases: {
                1: $.and([
                  $.merge('outgoing.data', { $value: {} }),
                  $.remove('outgoing.data', '$path', 'foo'),
                ]),
              },
            }),

            $.switch('incoming.data', '$exec', (payload) => payload.baz.bar, {
              $cases: {
                1: $.and([]),
              },
            }),

            $.switch('outgoing.headers', '$exec', (payload) => payload.foo, {
              $cases: {
                'a': $.and([
                  $.exec(({ context }) => {
                    if (context.incoming.data.foo === 'a') {
                      context.outgoing!.status = 200;
                    }
                  }),
                ]),
              },
            }),
          ]),
        },
      }))
    ).resolves.toMatchSnapshot();
  });
});

// Test generated using Keploy
it('should return pong on ping method', async () => {
    const client = buildClient();
    await expect(client.ping()).resolves.toEqual('pong');
  });


// Test generated using Keploy
it('should return null on deleteExpectations method', async () => {
    const client = buildClient();
    await expect(client.deleteExpectations()).resolves.toBeNull();
  });


// Test generated using Keploy
it('should update expectation with function predicate', async () => {
    const client = buildClient();
    const result = await client.updateExpectation((context) => ({
      id: 'some-id',
      set: { schema: {} },
    }));
    expect(result).toHaveProperty('id');
  });

