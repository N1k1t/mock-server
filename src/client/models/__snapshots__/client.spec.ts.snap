// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client.Models.Client should create expectation with complex schema as function 1`] = `
{
  "id": "12426262-85a3-4340-969b-272e003722e9",
  "isEnabled": true,
  "meta": {
    "additional": {},
    "executionsCount": 0,
  },
  "name": "MaliciousCobra",
  "schema": {
    "request": {
      "$and": [
        {
          "$has": {
            "$location": "path",
            "$value": "/foo",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$valueAnyOf": [
              "/foo",
              "/foo/bar",
            ],
          },
        },
        {
          "$has": {
            "$location": "path",
            "$match": "/foo/*",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$matchAnyOf": [
              "/foo/*",
              "*/bar",
            ],
          },
        },
        {
          "$has": {
            "$location": "path",
            "$regExp": {
              "flags": "",
              "source": "^\\/foo",
            },
          },
        },
        {
          "$has": {
            "$location": "path",
            "$regExpAnyOf": [
              {
                "flags": "",
                "source": "^\\/foo$",
              },
              {
                "flags": "",
                "source": "^\\/foo\\/bar$",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload.includes('/foo')",
            "$location": "path",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$path": "/foo",
            "$value": "/foo",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$path": "/foo",
            "$valueAnyOf": [
              "/foo",
              "/foo/bar",
            ],
          },
        },
        {
          "$has": {
            "$location": "path",
            "$match": "/foo/*",
            "$path": "/foo",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$matchAnyOf": [
              "/foo/*",
              "*/bar",
            ],
            "$path": "/foo",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$path": "/foo",
            "$regExp": {
              "flags": "",
              "source": "^\\/foo",
            },
          },
        },
        {
          "$has": {
            "$location": "path",
            "$path": "/foo",
            "$regExpAnyOf": [
              {
                "flags": "",
                "source": "^\\/foo$",
              },
              {
                "flags": "",
                "source": "^\\/foo\\/bar$",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload.includes('/foo')",
            "$location": "path",
            "$path": "/foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$value": {
              "bar": "a",
              "foo": "a",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$valueAnyOf": [
              {
                "bar": "a",
                "foo": "a",
              },
              {
                "bar": "b",
                "foo": "b",
              },
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$match": {
              "foo": "a",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$matchAnyOf": [
              {
                "foo": "a",
              },
              {
                "foo": "b",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload?.foo === 'a'",
            "$location": "incoming.query",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
            "$value": "a",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
            "$valueAnyOf": [
              "a",
              "b",
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$match": "a",
            "$path": "foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$matchAnyOf": [
              "a",
              "b",
            ],
            "$path": "foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
            "$regExp": {
              "flags": "",
              "source": "a|b",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
            "$regExpAnyOf": [
              {
                "flags": "i",
                "source": "a",
              },
              {
                "flags": "i",
                "source": "b",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload === 'a'",
            "$location": "incoming.query",
            "$path": "foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$value": {
              "bar": [
                "a",
              ],
              "foo": "a",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$valueAnyOf": [
              {
                "bar": [
                  "a",
                ],
                "foo": "a",
              },
              {
                "bar": [
                  "b",
                ],
              },
              {
                "baz": {},
              },
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$match": {
              "baz": {
                "foo": 1,
              },
              "foo": "a",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$matchAnyOf": [
              {
                "baz": {},
              },
              {
                "bar": [],
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload?.baz?.foo === 1",
            "$location": "incoming.body",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz",
            "$value": {
              "bar": 1,
              "foo": 1,
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz",
            "$valueAnyOf": [
              {
                "foo": 1,
              },
              {
                "bar": 1,
                "foo": 1,
              },
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$match": {
              "foo": 1,
            },
            "$path": "baz",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$matchAnyOf": [
              {},
              {
                "foo": 1,
              },
            ],
            "$path": "baz",
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload?.foo === 1",
            "$location": "incoming.body",
            "$path": "baz",
          },
        },
        {
          "$has": {
            "$jsonPath": "123.123.213",
            "$location": "incoming.body",
            "$match": {},
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz.baz",
            "$value": [
              1,
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz.baz",
            "$valueAnyOf": [
              [
                1,
              ],
              [
                1,
                2,
              ],
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$match": "*",
            "$path": "baz.baz.0",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$matchAnyOf": [
              "1",
              "2",
            ],
            "$path": "baz.baz.0",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz.baz",
            "$regExp": {
              "flags": "",
              "source": "1",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz.baz",
            "$regExpAnyOf": [
              {
                "flags": "",
                "source": "1",
              },
              {
                "flags": "",
                "source": "2",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload?.[0] === 1",
            "$location": "incoming.body",
            "$path": "baz.baz",
          },
        },
      ],
    },
    "response": {
      "$or": [
        {
          "$switch": {
            "$cases": {
              "200": {
                "$and": [
                  {
                    "$set": {
                      "$location": "outgoing.data",
                      "$value": {},
                    },
                  },
                  {
                    "$set": {
                      "$location": "outgoing.data",
                      "$path": "foo",
                      "$value": "a",
                    },
                  },
                  {
                    "$set": {
                      "$exec": "() => ({})",
                      "$location": "outgoing.headers",
                    },
                  },
                  {
                    "$set": {
                      "$location": "outgoing.headers",
                      "$path": "foo",
                      "$value": "bar",
                    },
                  },
                ],
              },
            },
            "$default": {
              "$set": {
                "$location": "outgoing.data",
                "$value": {},
              },
            },
            "$location": "outgoing.status",
          },
        },
        {
          "$switch": {
            "$cases": {
              "1": {
                "$and": [
                  {
                    "$merge": {
                      "$location": "outgoing.data",
                      "$value": {},
                    },
                  },
                  {
                    "$remove": {
                      "$location": "outgoing.data",
                      "$path": "foo",
                    },
                  },
                ],
              },
            },
            "$location": "incoming.body",
            "$path": "baz.bar",
          },
        },
        {
          "$switch": {
            "$cases": {
              "1": {
                "$and": [],
              },
            },
            "$exec": "(payload) => payload.baz.bar",
            "$location": "incoming.body",
          },
        },
        {
          "$switch": {
            "$cases": {
              "a": {
                "$and": [
                  {
                    "$exec": "({ context }) => {
                                    if (context.incoming.body.foo === 'a') {
                                        context.outgoing.status = 200;
                                    }
                                }",
                  },
                ],
              },
            },
            "$exec": "(payload) => payload.foo",
            "$location": "outgoing.headers",
          },
        },
      ],
    },
  },
  "type": "HTTP",
}
`;

exports[`Client.Models.Client should create expectation with complex schema as object 1`] = `
{
  "id": "12426262-85a3-4340-969b-272e003722e9",
  "isEnabled": true,
  "meta": {
    "additional": {},
    "executionsCount": 0,
  },
  "name": "MaliciousCobra",
  "schema": {
    "request": {
      "$and": [
        {
          "$has": {
            "$location": "path",
            "$value": "/foo",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$valueAnyOf": [
              "/foo",
              "/foo/bar",
            ],
          },
        },
        {
          "$has": {
            "$location": "path",
            "$match": "/foo/*",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$matchAnyOf": [
              "/foo/*",
              "*/bar",
            ],
          },
        },
        {
          "$has": {
            "$location": "path",
            "$regExp": {
              "flags": "",
              "source": "^\\/foo",
            },
          },
        },
        {
          "$has": {
            "$location": "path",
            "$regExpAnyOf": [
              {
                "flags": "",
                "source": "^\\/foo$",
              },
              {
                "flags": "",
                "source": "^\\/foo\\/bar$",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload.includes('/foo')",
            "$location": "path",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$path": "/foo",
            "$value": "/foo",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$path": "/foo",
            "$valueAnyOf": [
              "/foo",
              "/foo/bar",
            ],
          },
        },
        {
          "$has": {
            "$location": "path",
            "$match": "/foo/*",
            "$path": "/foo",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$matchAnyOf": [
              "/foo/*",
              "*/bar",
            ],
            "$path": "/foo",
          },
        },
        {
          "$has": {
            "$location": "path",
            "$path": "/foo",
            "$regExp": {
              "flags": "",
              "source": "^\\/foo",
            },
          },
        },
        {
          "$has": {
            "$location": "path",
            "$path": "/foo",
            "$regExpAnyOf": [
              {
                "flags": "",
                "source": "^\\/foo$",
              },
              {
                "flags": "",
                "source": "^\\/foo\\/bar$",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload.includes('/foo')",
            "$location": "path",
            "$path": "/foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$value": {
              "foo": "a",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$valueAnyOf": [
              {
                "foo": "a",
              },
              {
                "foo": "b",
              },
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$match": {
              "foo": "a",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$matchAnyOf": [
              {
                "foo": "a",
              },
              {
                "foo": "b",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload, { T }) => T(payload)?.foo === 'a'",
            "$location": "incoming.query",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
            "$value": "a",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
            "$valueAnyOf": [
              "a",
              "b",
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$match": "a",
            "$path": "foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$matchAnyOf": [
              "a",
              "b",
            ],
            "$path": "foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
            "$regExp": {
              "flags": "",
              "source": "a|b",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.query",
            "$path": "foo",
            "$regExpAnyOf": [
              {
                "flags": "i",
                "source": "a",
              },
              {
                "flags": "i",
                "source": "b",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload === 'a'",
            "$location": "incoming.query",
            "$path": "foo",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$value": {
              "bar": [
                "a",
              ],
              "foo": "a",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$valueAnyOf": [
              {
                "bar": [
                  "a",
                ],
                "foo": "a",
              },
              {
                "bar": [
                  "b",
                ],
              },
              {
                "baz": {},
              },
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$match": {
              "baz": {
                "foo": 1,
              },
              "foo": "a",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$matchAnyOf": [
              {
                "baz": {},
              },
              {
                "bar": [],
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload, { T }) => T(payload)?.baz?.foo === 1",
            "$location": "incoming.body",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz",
            "$value": {
              "bar": 1,
              "foo": 1,
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz",
            "$valueAnyOf": [
              {
                "foo": 1,
              },
              {
                "bar": 1,
              },
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$match": {
              "foo": 1,
            },
            "$path": "baz",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$matchAnyOf": [
              {},
              {
                "foo": "1",
              },
            ],
            "$path": "baz",
          },
        },
        {
          "$has": {
            "$exec": "(payload, { T }) => T(payload)?.foo === 1",
            "$location": "incoming.body",
            "$path": "baz",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz.baz",
            "$value": [
              1,
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz.baz",
            "$valueAnyOf": [
              [
                1,
              ],
              [
                1,
                2,
              ],
            ],
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$match": "*",
            "$path": "baz.baz",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$matchAnyOf": [
              "1",
              "2",
            ],
            "$path": "baz.baz",
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz.baz",
            "$regExp": {
              "flags": "",
              "source": "1",
            },
          },
        },
        {
          "$has": {
            "$location": "incoming.body",
            "$path": "baz.baz",
            "$regExpAnyOf": [
              {
                "flags": "",
                "source": "1",
              },
              {
                "flags": "",
                "source": "2",
              },
            ],
          },
        },
        {
          "$has": {
            "$exec": "(payload) => payload === 1",
            "$location": "incoming.body",
            "$path": "baz.baz",
          },
        },
      ],
    },
    "response": {
      "$or": [
        {
          "$switch": {
            "$cases": {
              "200": {
                "$and": [
                  {
                    "$set": {
                      "$location": "outgoing.data",
                      "$value": {},
                    },
                  },
                  {
                    "$set": {
                      "$location": "outgoing.data",
                      "$path": "foo",
                      "$value": "a",
                    },
                  },
                  {
                    "$set": {
                      "$exec": "() => ({})",
                      "$location": "outgoing.headers",
                    },
                  },
                  {
                    "$set": {
                      "$location": "outgoing.headers",
                      "$path": "foo",
                      "$value": "bar",
                    },
                  },
                ],
              },
            },
            "$default": {
              "$set": {
                "$location": "outgoing.data",
                "$value": {},
              },
            },
            "$location": "outgoing.status",
          },
        },
        {
          "$switch": {
            "$cases": {
              "1": {
                "$and": [
                  {
                    "$merge": {
                      "$location": "outgoing.data",
                      "$value": {},
                    },
                  },
                  {
                    "$remove": {
                      "$location": "outgoing.data",
                      "$path": "foo",
                    },
                  },
                ],
              },
            },
            "$location": "outgoing.data",
            "$path": "baz.bar",
          },
        },
        {
          "$switch": {
            "$cases": {
              "a": {
                "$and": [
                  {
                    "$exec": "({ context }) => {
                                                    if (context.incoming.body?.foo === 'a') {
                                                        context.outgoing.status = 200;
                                                    }
                                                }",
                  },
                ],
              },
            },
            "$exec": "(payload) => payload.foo",
            "$location": "outgoing.headers",
          },
        },
      ],
    },
  },
  "type": "HTTP",
}
`;
